test/integration/provider/alicloud.go

```diff
@@ -0,0 +1,174 @@
+
+	providerDriver "github.com/gardener/machine-controller-manager-provider-alicloud/pkg/alicloud"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/json"
+	"log"
+)
+
+/**
+	Orphaned Resources
+	- VMs:
+		Describe instances with specified tag name:<cluster-name>
+		Report/Print out instances found
+		Describe volumes attached to the instance (using instance id)
+		Report/Print out volumes found
+		Delete attached volumes found
+		Terminate instances found
+	- Disks:
+		Describe volumes with tag status:available
+		Report/Print out volumes found
+		Delete identified volumes
+**/
+
+func newSession(machineClass *v1alpha1.MachineClass, secret *v1.Secret) spi.ECSClient {
+	var (
+		providerSpec *api.ProviderSpec
+		sPI          spi.PluginSPIImpl
+	)
+
+	err := json.Unmarshal([]byte(machineClass.ProviderSpec.Raw), &providerSpec)
+	if err != nil {
+		providerSpec = nil
+		log.Printf("Error occured while performing unmarshal %s", err.Error())
+	}
+	sess, err := sPI.NewECSClient(secret, providerSpec.Region)
+	if err != nil {
+		log.Printf("Error occured while creating new session %s", err)
+	}
+	return sess
+}
+
+func getMachines(machineClass *v1alpha1.MachineClass, secretData map[string][]byte) ([]string, error) {
+	var machines []string
+	var sPI spi.PluginSPIImpl
+	driverProvider := providerDriver.NewAlicloudPlugin(&sPI)
+	machineList, err := driverProvider.ListMachines(context.TODO(), &driver.ListMachinesRequest{
+		MachineClass: machineClass,
+		Secret:       &v1.Secret{Data: secretData},
+	})
+	if err != nil {
+		return nil, err
+	} else if len(machineList.MachineList) != 0 {
+		for _, machine := range machineList.MachineList {
+			machines = append(machines, machine)
+		}
+	}
+	return machines, nil
+}
+
+// getOrphanesInstances returns list of Orphan resources that couldn't be deleted
+func getOrphanedInstances(tagName string, tagValue string, machineClass *v1alpha1.MachineClass, secretData map[string][]byte) ([]string, error) {
+	sess := newSession(machineClass, &v1.Secret{Data: secretData})
+	var instancesID []string
+	var tags = &[]ecs.DescribeInstancesTag{{Key: tagName, Value: tagValue}}
+	input := ecs.CreateDescribeInstancesRequest()
+	input.InstanceName = "instance-state-name"
+	input.Status = "running"
+	input.Tag = tags
+
+	result, err := sess.DescribeInstances(input)
+	if err != nil {
+		return instancesID, err
+	}
+	for _, instance := range result.Instances.Instance {
+		instancesID = append(instancesID, instance.InstanceId)
+	}
+	return instancesID, nil
+}
+
+// getOrphanesDisks returns list of Orphan disks
+func getOrphanedDisks(tagName string, tagValue string, machineClass *v1alpha1.MachineClass, secretData map[string][]byte) ([]string, error) {
+	sess := newSession(machineClass, &v1.Secret{Data: secretData})
+	var volumeID []string
+	var tags = &[]ecs.DescribeDisksTag{{Key: tagName, Value: tagValue}}
+	input := ecs.CreateDescribeDisksRequest()
+	input.Tag = tags
```

Comment:
search for disks which are `available` state

Reply:
done